File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        def helper (pre: int, inorderStart: int, inorderEnd: int) -> Optional[TreeNode]:
            if inorderStart > inorderEnd:
                return None
            
            inorderIndex = inorderStart
            while inorder[inorderIndex] != preorder[pre]:
                inorderIndex += 1

            head = TreeNode(preorder[pre])
            head.left = helper(pre + 1, inorderStart, inorderIndex - 1)
            head.right = helper(pre + inorderIndex - inorderStart + 1, inorderIndex + 1, inorderEnd)

            return head

        return helper(0, 0, len(inorder) - 1)

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        int cnt = 0;
        stack<TreeNode*> s;
        TreeNode* curr = root;

        while (curr || !s.empty()) {
            while (curr) {
                s.push(curr);
                curr = curr->left;
            }
            curr = s.top();
            s.pop();
            cnt++;
            if (cnt == k) return curr->val;
            curr = curr->right;

        }
    }
};
